core/src/test/java/org/infinispan/persistence/file/SingleFileGracefulShutdownMigrationTest.java

```diff
@@ -0,0 +1,159 @@
+ *          IntStream.range(1000, 2000).forEach(i -> cache.put(i, i, 1, TimeUnit.MINUTES));
+ *
+ *          // WrappedByteArrays
+ *          WrappedByteArray wba = new WrappedByteArray("wrapped-bytes".getBytes(StandardCharsets.UTF_8));
+ *          cache.put(wba, wba);
+ *
+ *          // Async Xsite PrivateMetadata entry
+ *          DefaultIracVersionGenerator iracVersionGenerator = new DefaultIracVersionGenerator("");
+ *          TestingUtil.replaceField("site-name", "localSite", iracVersionGenerator, DefaultIracVersionGenerator.class);
+ *          PrivateMetadata privateMetadata = new PrivateMetadata.Builder()
+ *                .iracMetadata(iracVersionGenerator.generateNewMetadata(2))
+ *                .build();
+ *          MarshallableEntry<Object, Object> me = mef.create("irac-key", "irac-value", null, privateMetadata, -1, -1);
+ *          store.write(me);
+ *
+ *          // Optimistic Tx PrivateMetadata entry
+ *          privateMetadata = new PrivateMetadata.Builder()
+ *                .entryVersion(new NumericVersionGenerator().generateNew())
+ *                .build();
+ *          me = mef.create("opt-tx-key", "opt-tx-value", null, privateMetadata, -1, -1);
+ *          store.write(me);
+ *       }
+ *    }
+ * </pre>
+ * <p>
+ * Files stored in `sfs/corrupt` contain already corrupted .dat files that were originally created in 11.x and were
+ * migrated to 12.1.x with the broken SFS migration code. Hence, when these .dat files are loaded the SFS attempts to
+ * recover the corrupt data.
+ *
+ * @author Ryan Emerson
+ * @since 13.0
+ */
+@Test(groups = "unit", testName = "persistence.file.SingleFileGracefulShutdownMigrationTest")
+public class SingleFileGracefulShutdownMigrationTest extends AbstractInfinispanTest {
+
+   private static final String CACHE_NAME = "update-cache";
+
+   private String tmpDirectory;
+
+   @BeforeClass
+   protected void setUpTempDir() throws IOException {
+      tmpDirectory = CommonsTestingUtil.tmpDirectory(this.getClass());
+      new File(tmpDirectory).mkdirs();
+   }
+
+   @AfterClass
+   protected void clearTempDir() {
+      Util.recursiveFileRemove(tmpDirectory);
+   }
+
+   enum Marshaller {
+      PROTOSTREAM,
+      JAVA_SERIALIZATION
+   }
+
+   @DataProvider(name = "testFiles")
+   Object[][] singleTypes() {
+      return new Object[][]{
+            // Not possible to test corrupt Marshaller.JAVA_SERIALIZATION bytes as previously 11.x -> 12. migration always fails with non-protostream marshallers
+            // MAGIC_11_0
+            {"sfs/11_0/all-entries-java-serialization.dat", Marshaller.JAVA_SERIALIZATION, 1003},
+            // MAGIC_12_0 created on 12.1.x
+            {"sfs/12_0/all-entries-java-serialization.dat", Marshaller.JAVA_SERIALIZATION, 1003},
+            // 11.x -> 12.1.x migration resulting in corrupt data
+            {"sfs/corrupt/all-entries.dat", Marshaller.PROTOSTREAM, 2003},
```

Comment:
Why do we have an extra 1000 entries? Did the corruption create new ones?

Reply:
It's due to recreating the lost `created` and `lastUsed` timestamps. In the other stores the values have expired so they aren't iterated over.

Reply:
Makes sense. Is there a place we should mention that expired entries may be resurrected when hitting this migration issue somewhere?

Reply:
I'm going to create a small doc describing the original problem as well as the various limitations of the recovery. Downstream we will need this for support. I guess I could put a condensed version of this in the upgrade notes?

Reply:
Yeah, I am not sure exactly where it should go upstream. But I guess that is as good as a spot as any. Just make sure to list the previously affected versions explicitly.